grpc/src/main/java/com/linecorp/armeria/common/grpc/GsonGrpcJsonMarshallerBuilder.java

```diff
@@ -30,42 +34,117 @@
+            final Function<JsonFormat.Parser, JsonFormat.Parser> cast =
+                    (Function<JsonFormat.Parser, JsonFormat.Parser>) jsonParserCustomizer;
+            this.jsonParserCustomizer = cast;
+        } else {
+            this.jsonParserCustomizer = this.jsonParserCustomizer.andThen(jsonParserCustomizer);
+        }
+        return this;
+    }
+
     /**
      * Adds a {@link Consumer} that can customize the {@link JsonFormat.Parser}
      * used when deserializing a JSON payload into a {@link Message}.
+     *
+     * @deprecated {@link JsonFormat.Parser} is immutable so all changes applied in the {@link Consumer}
+     *     will be lost. Please use the {@link #jsonParserCustomizer(Function) jsonParserCustomizer}
+     *     which accepts {@link Function} parameter instead.
      */
+    @Deprecated
     public GsonGrpcJsonMarshallerBuilder jsonParserCustomizer(
             Consumer<? super JsonFormat.Parser> jsonParserCustomizer) {
+        if (!loggedJsonParserCustomizerWarning) {
+            logger.warn("{}.jsonParserCustomizer(Consumer) does not work as expected, " +
+                            "use jsonParserCustomizer(Function).",
+                    getClass().getSimpleName());
+            loggedJsonParserCustomizerWarning = true;
+        }
         requireNonNull(jsonParserCustomizer, "jsonParserCustomizer");
         if (this.jsonParserCustomizer == null) {
             @SuppressWarnings("unchecked")
             final Consumer<JsonFormat.Parser> cast = (Consumer<JsonFormat.Parser>) jsonParserCustomizer;
-            this.jsonParserCustomizer = cast;
+            this.jsonParserCustomizer = parser -> {
+                cast.accept(parser);
+                return parser;
+            };
         } else {
-            this.jsonParserCustomizer = this.jsonParserCustomizer.andThen(jsonParserCustomizer);
+            this.jsonParserCustomizer = this.jsonParserCustomizer.andThen(parser -> {
+                jsonParserCustomizer.accept(parser);
+                return parser;
+            });
         }
         return this;
     }
 
     /**
      * Adds a {@link Consumer} that can customize the {@link JsonFormat.Printer}
      * used when serializing a {@link Message} into a JSON payload.
+     *
+     * @deprecated {@link JsonFormat.Printer} is immutable so all changes applied in the {@link Consumer}
+     *     will be lost. Please use the {@link #jsonPrinterCustomizer(Function) jsonParserCustomizer}
+     *     which accepts {@link Function} parameter instead.
      */
+    @Deprecated
     public GsonGrpcJsonMarshallerBuilder jsonPrinterCustomizer(
             Consumer<? super JsonFormat.Printer> jsonPrinterCustomizer) {
+        if (!loggedJsonPrinterCustomizerWarning) {
+            logger.warn("{}.jsonPrinterCustomizer(Consumer) does not work as expected," +
+                            "use jsonPrinterCustomizer(Function).",
+                    getClass().getSimpleName());
```

Comment:
Let me fix the indentation:
```suggestion
            logger.warn("{}.jsonPrinterCustomizer(Consumer) does not work as expected; " +
                        "use jsonPrinterCustomizer(Function).",
                        getClass().getSimpleName());
```

Reply:
Thanks! Intellij haven't formated it and there was no linter error so I assumed that indentation is ok 😃

Reply:
No worries. 😄